keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1806 @@
+2. Pod Level Resource Specifications is an **opt-in** feature, and will have no
+effect on existing deployments. Only deployments that explicitly require this
+functionality should turn it on by setting the relevant resource section at
+pod-level in the Pod specification.
+
+### Risks and Mitigations
+
+1. Since Pod Level Resource Specifications is an **opt-in** feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+
+2. While this feature doesn't alter the existing cgroups structure, it does
+   change how pod-level cgroup values are determined. Currently, Kubernetes
+   derived these values from the container-level cgroup settings. However, with
+   Pod Level Resource Specifications enabled, pod-level cgroup settings will be
+   directly set based on the values specified in the pod's resource spec stanza,
+   if set. This change in behavior could potentially affect:
+
+    * Workloads or tools that rely on reading cgroup values: This means that any
+   workloads or tools that depend on reading or interpreting container cgroup
+   values might observe different derived values if pod-level resources are
+   specified without container level settings.
+
+    * Third-party schedulers or tools that make assumptions about pod-level
+      resource calculation: These tools might require adjustments to accommodate
+      the new way pod-level resources are determined. 
+   
+   To mitigate potential issues, the feature documentation will clearly highlight
+   this change and its potential impact. This will allow users to:
+    * Adjust their pod-level and container-level resource settings as needed
+    * Modify any custom schedulers or tools to align with new resource
+      calculation method.
+
+## Design Details
+
+### Components/Features changes
+
+#### Cgroup Structure Remains unchanged
+
+This KEP introduces pod-level resource specifications but does not alter the
+fundamental cgroup structure. Currently, pod level cgroups are derived from the
+resource requests and limits defined at the container level in the PodSpec. With
+this KEP, the PodSpec is extended to include pod-level resource requests and
+limits. These values will be directly used to configure the pod-level cgroup,
+providing a more direct and explicit way to manage the overall resource
+consumption of the pod.
+
+#### PodSpec API changes
+
+New field in `PodSpec`
+
+```
+type PodSpec struct {
+...
+  // The total amount of compute resources required by all containers in the pod.
+  // Currently, only CPU and Memory resources are supported .
+  // +featureGate=PodLevelResources
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+#### PodSpec Validation Rules
+
+##### Proposed Validation Rules
```

Comment:
I'm unsure if it's specified anywhere in this KEP, but I couldn't find it.
What happens if multiple containers specify their resources, and the sum of those resources is greater than the Pod resources specified?

Reply:
I found it:
>  **Rule: Aggregated container requests cannot be greater than pod-level
  request or limit.**
> * Aggregate container-level limits can be greater than pod-level limits, but the
  combined resource consumption of all the containers within a pod will always be
  less than the limit.